File: backend/app/schemas/overlay_template.py
"""
覆盖层模板相关 Schemas
"""
from typing import Optional, List, Literal, Any, Dict
from pydantic import BaseModel, Field, model_validator
from datetime import datetime


# ============================================================================
# 锚点类型
# ============================================================================

ContainerAnchor = Literal[
    'top-left', 'top', 'top-right',
    'left', 'center', 'right',
    'bottom-left', 'bottom', 'bottom-right'
]

ElementAnchor = ContainerAnchor  # 元素锚点与容器锚点相同

# ============================================================================
# 数据源类型
# ============================================================================

DataSource = Literal[
    'province',           # 省份
    'city',               # 城市
    'district',           # 区县
    'province_en',        # 省份英文名
    'city_en',            # 城市英文名
    'district_en',        # 区县英文名
    'region',             # 区域（组合显示，如"河南省 三门峡市 湖滨区"）
    'region_en',          # 区域英文（组合显示，如"Hubin District, Sanmenxia City, Henan Province"）
    'road_number',        # 道路编号
    'road_name',          # 道路名称
    'road_name_en',       # 道路英文名
    'speed',              # 速度
    'elevation',          # 海拔
    'compass_angle',      # 方位角
    'elapsed_distance',   # 已行驶距离
    'elapsed_time',       # 已用时间
    'remain_distance',    # 剩余距离
    'remain_time',        # 剩余时间
    'current_time',       # 当前时间
    'latitude',           # 纬度
    'longitude',          # 经度
]

# ============================================================================
# 位置配置
# ============================================================================

class PositionConfig(BaseModel):
    """位置配置"""
    container_anchor: ContainerAnchor = Field(default='top-left', description="容器锚点")
    element_anchor: ElementAnchor = Field(default='top-left', description="元素锚点")
    x: float = Field(default=0.0, ge=-0.5, le=0.5, description="X偏移（画布宽度的百分比）")
    y: float = Field(default=0.0, ge=-0.5, le=0.5, description="Y偏移（画布高度的百分比）")
    use_safe_area: bool = Field(default=False, description="是否使用安全区")

# ============================================================================
# 尺寸配置
# ============================================================================

class SizeConfig(BaseModel):
    """尺寸配置（百分比）"""
    width: Optional[float] = Field(None, ge=0, le=1, description="宽度（画布宽度的百分比）")
    height: Optional[float] = Field(None, ge=0, le=1, description="高度（画布高度的百分比）")

# ============================================================================
# 内容配置
# ============================================================================

class ContentConfig(BaseModel):
    """内容配置"""
    source: DataSource = Field(description="数据来源")
    prefix: str = Field(default='', description="前缀（已废弃，使用format）")
    suffix: str = Field(default='', description="后缀（已废弃，使用format）")
    format: str = Field(default='{}', description="格式化字符串，支持前缀后缀和小数位数，如'速度: {:.1f} km/h'")
    sample_text: Optional[str] = Field(None, description="自定义示例文本（预览时使用）")
    decimal_places: Optional[int] = Field(None, ge=0, le=10, description="小数位数（仅数字类型）")

# ============================================================================
# 文本布局配置
# ============================================================================

class TextLayoutConfig(BaseModel):
    """文本布局配置"""
    width: Optional[float] = Field(None, ge=0, le=1, description="固定宽度")
    max_width: Optional[float] = Field(None, ge=0, le=1, description="最大宽度")
    min_width: Optional[float] = Field(None, ge=0, le=1, description="最小宽度")
    height: Optional[float] = Field(None, ge=0, le=1, description="固定高度")
    horizontal_align: Literal['left', 'center', 'right', 'justify'] = Field(default='left', description="水平对齐")
    vertical_align: Literal['top', 'middle', 'bottom'] = Field(default='top', description="垂直对齐")
    wrap: bool = Field(default=False, description="是否折行")
    max_lines: Optional[int] = Field(None, ge=1, description="最大行数")
    line_height: float = Field(default=1.2, ge=0.5, le=3.0, description="行高倍数")
    char_spacing: Optional[float] = Field(None, ge=0, description="字间距（相对于字号的百分比）")
    paragraph_spacing: Optional[float] = Field(None, ge=0, description="段落间距（相对于字号的百分比）")

# ============================================================================
# 样式配置
# ============================================================================

class StyleConfig(BaseModel):
    """样式配置"""
    font_family: str = Field(default='system_msyh', description="字体ID")
    font_size: float = Field(default=0.03, gt=0, le=0.2, description="字体大小（画布高度的百分比）")
    color: str = Field(default='#FFFFFF', description="文本颜色")
    background_color: Optional[str] = Field(None, description="背景颜色")
    padding: Optional[float] = Field(None, ge=0, description="内边距（相对于字号）")

# ============================================================================
# 元素配置
# ============================================================================

class OverlayElementBase(BaseModel):
    """覆盖层元素基础配置"""
    id: str = Field(description="元素ID")
    type: Literal['text', 'road_sign', 'icon', 'group'] = Field(default='text', description="元素类型")
    name: str = Field(description="元素名称")
    visible: bool = Field(default=True, description="是否可见")

    # 位置和尺寸
    position: PositionConfig = Field(default_factory=PositionConfig, description="位置配置")
    size: SizeConfig = Field(default_factory=SizeConfig, description="尺寸配置")

    # 样式（所有元素通用）
    style: Optional[StyleConfig] = Field(default=None, description="样式配置")

    # 数据绑定（非group元素）
    content: Optional[ContentConfig] = Field(default=None, description="内容配置")

    # 文本布局（仅text元素）
    layout: Optional[TextLayoutConfig] = Field(default=None, description="文本布局配置")

    # 子元素（仅group元素）
    children: Optional[List['OverlayElement']] = Field(default=None, description="子元素")


# 支持自引用
OverlayElement = OverlayElementBase

# ============================================================================
# 安全区配置
# ============================================================================

class SafeAreaConfig(BaseModel):
    """安全区配置"""
    top: float = Field(default=0.05, ge=0, le=0.2, description="顶部边距")
    bottom: float = Field(default=0.05, ge=0, le=0.2, description="底部边距")
    left: float = Field(default=0.05, ge=0, le=0.2, description="左侧边距")
    right: float = Field(default=0.05, ge=0, le=0.2, description="右侧边距")

# ============================================================================
# 背景配置
# ============================================================================

class BackgroundConfig(BaseModel):
    """背景配置"""
    color: str = Field(default='#000000', description="背景颜色")
    opacity: float = Field(default=0.0, ge=0, le=1, description="不透明度")

# ============================================================================
# 模板配置
# ============================================================================

class CanvasConfig(BaseModel):
    """画布配置"""
    width: int = Field(default=1920, ge=100, le=7680, description="画布宽度（像素）")
    height: int = Field(default=1080, ge=100, le=4320, description="画布高度（像素）")


class OverlayTemplateConfig(BaseModel):
    """覆盖层模板配置"""
    canvas: CanvasConfig = Field(default_factory=CanvasConfig, description="画布配置")
    safe_area: SafeAreaConfig = Field(default_factory=SafeAreaConfig, description="安全区配置")
    background: Optional[BackgroundConfig] = Field(default=None, description="背景配置")
    elements: List[OverlayElement] = Field(default_factory=list, description="元素列表")

# ============================================================================
# API 请求/响应模型
# ============================================================================

class OverlayTemplateCreate(BaseModel):
    """创建覆盖层模板请求"""
    name: str = Field(..., min_length=1, max_length=100, description="模板名称")
    description: Optional[str] = Field(None, description="模板描述")
    config: OverlayTemplateConfig = Field(..., description="模板配置")


class OverlayTemplateUpdate(BaseModel):
    """更新覆盖层模板请求"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="模板名称")
    description: Optional[str] = Field(None, description="模板描述")
    config: Optional[OverlayTemplateConfig] = Field(None, description="模板配置")
    is_public: Optional[bool] = Field(None, description="是否公开")


class OverlayTemplateResponse(BaseModel):
    """覆盖层模板响应"""
    id: int
    name: str
    description: Optional[str]
    config: OverlayTemplateConfig
    user_id: Optional[int]
    is_public: bool
    is_system: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class OverlayTemplateListResponse(BaseModel):
    """覆盖层模板列表响应"""
    total: int
    items: List[OverlayTemplateResponse]


# ============================================================================
# 导入/导出
# ============================================================================

class OverlayTemplateImport(BaseModel):
    """导入覆盖层模板"""
    file: bytes  # YAML 文件内容


class OverlayTemplateExport(BaseModel):
    """导出覆盖层模板"""
    yaml_content: str


# ============================================================================
# 预览
# ============================================================================

class OverlayPreviewRequest(BaseModel):
    """预览请求"""
    template_id: int
    resolution: str = Field(default='1920x1080', description="预览分辨率")


# ============================================================================
# 字体相关
# ============================================================================

class FontCreate(BaseModel):
    """创建字体请求（管理员上传）"""
    name: str = Field(..., max_length=100, description="字体名称")


class FontResponse(BaseModel):
    """字体响应"""
    id: str
    name: str
    filename: str
    type: Literal['system', 'admin', 'user']
    owner_id: Optional[int]
    file_size: int
    family: Optional[str]
    style: str
    weight: int
    supports_latin: bool
    supports_chinese: bool
    supports_japanese: bool
    supports_korean: bool
    preview_url: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class FontListResponse(BaseModel):
    """字体列表响应"""
    total: int
    items: List[FontResponse]


# ============================================================================
# 导出配置
# ============================================================================

class OverlayExportRequest(BaseModel):
    """覆盖层导出请求（使用模板配置中的画布尺寸）"""
    template_id: int = Field(..., description="模板ID")
    frame_rate: int = Field(default=1, ge=0, description="采样率（0=全部，1=每秒1帧）")
    start_index: int = Field(default=0, ge=0, description="起始点索引")
    end_index: int = Field(default=-1, description="结束点索引（-1=到最后）")
    output_format: Literal['zip', 'png_sequence'] = Field(default='zip', description="输出格式")


class PreviewWithConfigRequest(BaseModel):
    """使用配置生成预览的请求"""
    config: OverlayTemplateConfig = Field(..., description="模板配置")

File: frontend/src/api/overlayTemplate.ts
/**
 * 覆盖层模板 API
 */
import request from './request'
import type { AxiosProgressEvent } from 'axios'

// ============================================================================
// 类型定义
// ============================================================================

export type ContainerAnchor =
  | 'top-left' | 'top' | 'top-right'
  | 'left' | 'center' | 'right'
  | 'bottom-left' | 'bottom' | 'bottom-right'

export type ElementAnchor = ContainerAnchor

export type DataSource =
  | 'none'  // 自定义文本，不从数据源获取
  | 'province' | 'city' | 'district'
  | 'province_en' | 'city_en' | 'district_en'
  | 'region' | 'region_en'
  | 'road_number' | 'road_name' | 'road_name_en'
  | 'speed' | 'elevation' | 'compass_angle'
  | 'elapsed_distance' | 'elapsed_time'
  | 'remain_distance' | 'remain_time'
  | 'current_time' | 'latitude' | 'longitude'

export interface PositionConfig {
  container_anchor: ContainerAnchor
  element_anchor: ElementAnchor
  x: number
  y: number
  use_safe_area: boolean
}

export interface SizeConfig {
  width?: number
  height?: number
}

export interface ContentConfig {
  source: DataSource
  prefix: string  // 已废弃，使用 format
  suffix: string  // 已废弃，使用 format
  format: string  // 支持前缀后缀和小数位数，如 '速度: {:.1f} km/h'
  sample_text?: string  // 自定义示例文本
  decimal_places?: number  // 小数位数（0-10）
}

export interface TextLayoutConfig {
  width?: number
  max_width?: number
  min_width?: number
  height?: number
  horizontal_align: 'left' | 'center' | 'right' | 'justify'
  vertical_align: 'top' | 'middle' | 'bottom'
  wrap: boolean
  max_lines?: number
  line_height: number
  char_spacing?: number
  paragraph_spacing?: number
}

export interface StyleConfig {
  font_family: string
  font_size: number
  color: string
  background_color?: string
  padding?: number
}

export interface OverlayElement {
  id: string
  type: 'text' | 'road_sign' | 'icon' | 'group'
  name: string
  visible: boolean | string
  position: PositionConfig
  size: SizeConfig
  style?: StyleConfig
  content?: ContentConfig
  layout?: TextLayoutConfig
  children?: OverlayElement[]
}

export interface SafeAreaConfig {
  top: number
  bottom: number
  left: number
  right: number
}

export interface BackgroundConfig {
  color: string
  opacity: number
}

export interface CanvasConfig {
  width: number
  height: number
}

export interface OverlayTemplateConfig {
  canvas: CanvasConfig
  safe_area: SafeAreaConfig
  background?: BackgroundConfig
  elements: OverlayElement[]
}

export interface OverlayTemplate {
  id: number
  name: string
  description: string | null
  config: OverlayTemplateConfig
  user_id: number | null
  is_public: boolean
  is_system: boolean
  created_at: string
  updated_at: string
}

export interface OverlayTemplateCreate {
  name: string
  description?: string
  config: OverlayTemplateConfig
}

export interface OverlayTemplateUpdate {
  name?: string
  description?: string
  config?: OverlayTemplateConfig
  is_public?: boolean
}

export interface OverlayTemplateListResponse {
  total: number
  items: OverlayTemplate[]
}

export interface Font {
  id: string
  name: string
  filename: string
  type: 'system' | 'admin' | 'user'
  owner_id: number | null
  file_size: number
  family: string | null
  style: string
  weight: number
  supports_latin: boolean
  supports_chinese: boolean
  supports_japanese: boolean
  supports_korean: boolean
  preview_url: string | null
  created_at: string
}

export interface FontListResponse {
  total: number
  items: Font[]
}

// ============================================================================
// API 方法
// ============================================================================

export const overlayTemplateApi = {
  /**
   * 获取模板列表
   */
  list: (params?: { include_system?: boolean; only_public?: boolean }) => {
    return request.get<OverlayTemplateListResponse>('/overlay-templates', { params })
  },

  /**
   * 获取单个模板
   */
  get: (templateId: number) => {
    return request.get<OverlayTemplate>(`/overlay-templates/${templateId}`)
  },

  /**
   * 创建模板
   */
  create: (data: OverlayTemplateCreate) => {
    return request.post<OverlayTemplate>('/overlay-templates', data)
  },

  /**
   * 更新模板
   */
  update: (templateId: number, data: OverlayTemplateUpdate) => {
    return request.put<OverlayTemplate>(`/overlay-templates/${templateId}`, data)
  },

  /**
   * 删除模板
   */
  delete: (templateId: number) => {
    return request.delete(`/overlay-templates/${templateId}`)
  },

  /**
   * 复制模板
   */
  duplicate: (templateId: number) => {
    return request.post<OverlayTemplate>(`/overlay-templates/${templateId}/duplicate`)
  },

  /**
   * 导出模板为 YAML
   */
  export: (templateId: number) => {
    return request.get<string>(`/overlay-templates/${templateId}/export`, {
      responseType: 'blob'
    })
  },

  /**
   * 从 YAML 导入模板
   */
  import: (file: File) => {
    const formData = new FormData()
    formData.append('file', file)
    return request.post<OverlayTemplate>('/overlay-templates/import', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    })
  },

  /**
   * 生成模板预览图（使用模板配置中的画布尺寸）
   */
  preview: (templateId: number) => {
    return request.get<Blob>(`/overlay-templates/${templateId}/preview`, {
      responseType: 'blob'
    })
  },

  /**
   * 使用指定配置生成预览图（不保存到数据库）
   */
  previewWithConfig: (config: OverlayTemplateConfig) => {
    return request.post<Blob>('/overlay-templates/preview-with-config', { config }, {
      responseType: 'blob'
    })
  },

  /**
   * 获取字体列表
   */
  listFonts: () => {
    return request.get<FontListResponse>('/overlay-templates/fonts/list')
  },

  /**
   * 上传字体
   */
  uploadFont: (file: File, onProgress?: (progress: number) => void) => {
    const formData = new FormData()
    formData.append('file', file)
    return request.post<Font>('/overlay-templates/fonts/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      onUploadProgress: (progressEvent: AxiosProgressEvent) => {
        if (onProgress && progressEvent.total) {
          const percentCompleted = Math.round((progressEvent.loaded * 100) / progressEvent.total)
          onProgress(percentCompleted)
        }
      }
    })
  },

  /**
   * 删除字体
   */
  deleteFont: (fontId: string) => {
    return request.delete(`/overlay-templates/fonts/${fontId}`)
  },

  /**
   * 导出覆盖层序列（使用模板配置中的画布尺寸）
   */
  exportOverlay: (
    trackId: number,
    params: {
      template_id: number
      frame_rate?: number
      start_index?: number
      end_index?: number
      output_format?: 'zip' | 'png_sequence'
    }
  ) => {
    const formData = new FormData()
    Object.entries(params).forEach(([key, value]) => {
      formData.append(key, String(value))
    })
    return request.post<Blob>(`/overlay-templates/tracks/${trackId}/export`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      responseType: 'blob'
    })
  }
}
